### PR TITLE
Problem: bulk_create can fail with IntegrityError

### PR DESCRIPTION
**Pull request by dkliban**
Solution: add a manager that provides a bulk_get_or_create() method

closes #4060
https://pulp.plan.io/issues/4060

### CODE REVIEW

File: pulpcore/pulpcore/app/models/content.py [hunk 1]
```diff
@@ -4,14 +4,64 @@
...
+        """
+        try:
+            return super().bulk_create(objs, batch_size=batch_size)
+        except IntegrityError:
+            objs = list(objs)
+            for i in range(len(objs)):
+                try:
+                    objs[i].save()
+                except IntegrityError:
+                    objs[i] = objs[i].objects.filter(objs[i].q())
```
Comment (gmbnomis) [inline pulpcore/pulpcore/app/models/content.py]: I didn't test this, but this looks like it should be a `get()`.

Comment (dkliban) [reply pulpcore/pulpcore/app/models/content.py]: Both work actually, but I did switch it to get().

**Comment by gmbnomis** [comment]
I like this a lot! For pulp_cookbook, it solves a problem similar to the Docker problem mentioned in the issue.

File: pulpcore/pulpcore/app/models/content.py [hunk 2]
```diff
@@ -4,14 +4,64 @@
...
 from pulpcore.app.models import Model, MasterModel, Notes, GenericKeyValueRelation, storage, fields
 from pulpcore.exceptions import DigestValidationError, SizeValidationError
 
 
-class Artifact(Model):
+class BulkCreateManager(models.Manager):
+    """
+    A manager that provides a bulk_get_or_create()
+    """
+    def bulk_get_or_create(self, objs, batch_size=None):
```
Comment (bmbouter) [inline pulpcore/pulpcore/app/models/content.py]: a nit, but can a line of whitespace be added above this  for readability?

File: pulpcore/pulpcore/app/models/content.py [hunk 3]
```diff
@@ -4,14 +4,64 @@
...
+                except ValueError or IntegrityError:
+                    objs[i] = objs[i].__class__.objects.get(objs[i].q())
+        return objs
+
+
+class QueryMixin():
+    """
+    A mixin that provides models with querying utilities
+    """
+    def q(self):
```
Comment (bmbouter) [inline pulpcore/pulpcore/app/models/content.py]: adding a line above here too if you don't mind

**Pull request merged**